### ISSUE TITLE
Support for manual content reprojection

### ISSUE THREAD

**Issue by mhevery** [angular/angular#1989]
# GOAL

- provide `<content>`-like way of re-projecting content from one location of the tree to another. 
- Preserve the injection and context variables of the source

## Implementation

Here is a proposed implementation skeleton.

* make `TemplateRef` queryable (it already is injectable)

### `ng-outlet`

```javascript
@Directive({
  selector: '[ng-outlet]',
  properties: {
    'ngOutlet': 'ngOutlet',
    'ngOutletLocals': 'ngOutletLocals'
  },
  lifecycle: [onChange]
})
class NgOutlet {
  ngOutletLocals:any;
  ngOutlet:TemplateRef; // Maybe this should take an array of templates???

  constructor(public viewContainer: ViewContainerRef) { }

  onChange() {
    // todo: Don't destroy the view if ngTemplateRef did not change
    this.viewContainer.clear();
    var viewRef:ViewRef = this.viewContainer.createEmbeddedView(this.ngOutlet);
    // use structural changes to cal setLocal only if hash `ngOutletLocals` changed.
    for(var prop in this.ngOutletLocals) {
      viewRef.setLocal(prop, this.ngOutletLocals[prop]);
    }
  }
}
```

## Example of Usage

```javascript
@Component({
  selector: 'my-component'
})
@View({
  template: `
  <button template="var i = index" (click)="onClick()">
    {{i}}: {{text}}
  </button>

  <!-- here for simplicity, but it could be anywhere in the tree -->
  <template *ng-outlet="templates.first; locals={'index': 1}"></template>
`
})
class MyComponent {
  text: string;
  templates: QuerList<TemplateRef>;

  constructor(@ViewQuery(TemplateRef) templates: QueryList<TemplateRef>) {
    this.templates = templates;
  }

  onClick() {
    // do something
  }
}
```

**Comment by matanlurey** [comment]
I like. I have a prototype of something similar we can look at when your'e back.

**Comment by tbosch** [comment]
We could extend this a little bit to implement a faster shadow dom emulation:

Allow a `name` attribute on both, `ng-outlet` and `ng-template`. If `name` is set in `ng-outlet`, it will look up the correct `ElementRef` and `ProtoViewRef` in the light dom of the current component. 

E.g.
```html
<my-component>
  <button ng-template-ref name="btn"/>
</my-component>
```

```javascript
@View({
  template: '<ng-outlet name="btn"/>'
})
@Component({
  selector: 'my-component'
})
class MyComponent {}
```

**Comment by tbosch** [comment]
The difference to real shadow dom emulation are then:

- this way is probably faster, as it using views directly and does not need to query the DOM
- a user can introduce new locals (see above), which regular shadow dom emulation cannot.
- does not support redistribution (e.g. a `ng-outlet` goes away and the nodes are redistributed to other `ng-outlets`). However, I don't know of good use cases for this

**Comment by tbosch** [comment]
Removed my comments about shadow dom emulation in favor of https://github.com/angular/angular/issues/2529

**Comment by tbosch** [comment]
Some changes to our API surface that we should do with this:

Previous:
```javascript
AppViewManager.createViewInContainer(viewContainerLocation: ElementRef, atIndex: number,
                        protoViewRef: ProtoViewRef, context: ElementRef = null,
                        injector: Injector = null): ViewRef { ... }

ViewContainer.create(protoViewRef: ProtoViewRef = null, atIndex: number = -1,
                        context: ElementRef = null,injector: Injector = null): ViewRef { ... }

```

These methods actually cover 2 use cases:

1. manual content projection by using a context `ElementRef`. In this case, the `ProtoViewRef` needs to be taken from the context `ElementRef`, and using an explicit `injector` does not make sense as this 
use case means "Show this template somewhere else, but use the execution context, variables and injector from here".
2. dynamic component loading: here, a `ProtoViewRef` is created via a call to `Compiler.compileInHost` and then passed to one of the above methods. In this case, there is never a context `ElementRef`, but most likely an `Injector`

A better API would be as follows:

```javascript
AppViewManager.createEmbeddedView(viewContainerLocation: ElementRef, atIndex: number,
                        context: ElementRef = null) { ... }
AppViewManager.createHostView(viewContainerLocation: ElementRef, atIndex: number,
                        protoViewRef: ProtoViewRef, injector: Injector = null) { ... }

ViewContainer.create(atIndex: number = -1,
                        context: ElementRef = null): ViewRef { ... }
ViewContainer.createHostView(protoViewRef: ProtoViewRef = null, atIndex: number = -1,
                        injector: Injector = null): ViewRef { ... }
```

For `createEmbeddedView` the `ProtoViewRef` would always be used from the `context`. This also simplifies the proposal above in that `TemplateRef` only needs an `ElementRef`.

**Comment by tbosch** [comment]
Needs to support the cases: SubView, single stamp out, multiple stamp out

**Comment by tbosch** [comment]
Depends on #2529 as this will introduce SubViews which should be transparent to the user of this API.

**Comment by tbosch** [comment]
I think we can solve this a little nicer:

1. `TemplateRef` is already a primitive and injectable. We should make it also queryable, so that we don't need an extra directive to mark templates
2. For using a `TemplateRef` we should extend `ng-if` and `ng-for` to take a `TemplateRef` + extra locals via a data binding.

**Comment by mhevery** [comment]
@tbosch What is the status of this? can this be closed?

**Comment by tbosch** [comment]
`TemplateRef` is in master and is also queriable. However, I didn't have the time to change `NgFor` to support custom templates, nor implement `ngOutlet`. I have a local wip branch for this...
Won't continue on this for the next 3 weeks or so as I will focus on the compiler improvements, and GT is happy so far...

**Comment by mhevery** [comment]
This is done.

**Comment by user414** [comment]
This seems to be exactly what I'm looking for except that the documentation for it is lacking. Looking at the code it seems NgOutlet never made it in but ngForTemplate is the one to use, seems possible to use this with one example I saw in some other issue which I will link here for reference

https://github.com/angular/angular/issues/4769

but would be nice to add this in the doc of the ngFor

https://angular.io/docs/ts/latest/guide/template-syntax.html#!#ngFor

if I come up with a simple plunker to show this will post it here.

**Comment by zoechi** [comment]
@user141 https://github.com/angular/angular.io/pull/544/files (search for `ngForTemplate`)

**Comment by user414** [comment]
@zoechi Thank you very much

**Comment by user414** [comment]
I didn't realize until now that the doc is in another project, I will always make sure in the future I look up the angular.io issues/merge request for info, thanks again @zoechi

**Comment by user414** [comment]
Since @zoechi pointer saved me time thought I would use some of that time to post back some comments. Overall the implementation was as expected and described in the new docs except that I would have expected the index to work in the called template like this
```
<template #hero #i="index">
<hero-detail [hero]="hero"><span>This is my index {{i}}</span></hero-detail>
</template>
```
Is it stands now if you need the index in the template pointed to in the ngForTemplate it seems you are out of luck, or at least couldn't make it work. 

I would also have expected, and maybe I missed how to do this, to be able to easily define a template somewhere outside for reuse. In the hero example I would have expected to be able to define the 
```
<template #hero #i="index">
```
completely outside of
```
<hero-dashboard [heroes]="heroes">
```
so that in app.component.ts I could obtain a templateRef somehow and bind it to the child component instead like this
```
<hero-dashboard [heroes]="heroes" [heroDetailTmpl]="templateRef">
```

so that I can reuse it in multiple child component if need be without copy pasting. Thanks.

**Comment by user414** [comment]
Will modify my last comment, the index actually works. In my example, I was using it with just one element hence why I didn't see it changing. Which brings me to another comments. It might be good to modify the example in the new documentation to add one with the index to show that it can indeed work.

**Comment by user414** [comment]
A few more comments. It would also be useful to document that it does not seem possible to pass variable/parameter other then built in index to ngForTemplate target like this

```
<template ngFor  [ngForOf]="[jsonObject]" [ngForTemplate]="customRenderer" #itemNumber="itemNumber"></template>

<template #customRender #itemNumber #parentObject>
   <span>{{whichScopeAmI(itemNumber)}}</span>
  <span>{{parentObject.whichScopeAmI(itemNumber)}}</span>
</template>
```
I found that it is not possible but even though I think it might be a good feature to add, it would also be good to document what works and doesn't work. I would expect this ngForTemplate to work similarly to a method call where I wouldn't need to use ngFor and could just call another template defined somewhere else with extra parameter. However, it doesn't work like that and I think should be documented. It would also be good to document the scope of this target template, through experimentation I came to the conclusion that the scope is where the template is declared but I'm not sure it would have been my vote, it would be nice to have a reference to the object where the template is being run so at least we can access method there, otherwise I find it looses some flexibility since you have no access to where you will run. All of this could be done in equivalent javascript/typescript function and inheritance but kind of defeats the purpose.
